### PR TITLE
918/setting device passcode

### PR DESCRIPTION
**Pull request by commjoen**
This is an initial PR for:
- #918 : Add MASVS 2.11 to iOS Chapter about Data Storage
- #935: MASVS 4.7 is missing in MSTG
- #936: MASVS 4.11 is missing in MSTG

Please review it, give feedback & I will prepare it for a second review & final merge.

### CODE REVIEW

File: Document/0x05f-Testing-Local-Authentication.md [hunk 1]
```diff
@@ -1,6 +1,100 @@
...
+                // operation. Go to error/cancellation flow.
+            }
+        }
+    }
+
+```
+#### Dynamic Analysis
+Patch the app or use runtime instrumentation to bypass fingerprint authentication on the client. For example, you could use Frida to call the `onActivityResult` callback method directly to see if the cryptographic material (e.g. the setup cipher) can be ignored to proceed with the local authentication flow. Refer to the chapter "Tampering and Reverse Engineering on Android" for more information.
+
+### References
```
Comment (sushi2k) [inline Document/0x05f-Testing-Local-Authentication.md]: All the references are summarised now at the very end of each chapter, not after the test case. Please merge with references at the end of chapter.

File: Document/0x06f-Testing-Local-Authentication.md [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 During local authentication, an app authenticates the user against credentials stored locally on the device. In other words, the user "unlocks" the app or some inner layer of functionality by providing a valid PIN, password, or fingerprint, verified by referencing local data. Generally, this done so that users can more conveniently resume an existing session with a remote service or as a means of step-up authentication to protect some critical function.
 
+As stated before in chapter 0x04e: the tester should be aware that local authentication should always be enforced at a remote endpoint or based on a cryptographic primitive. Attackers can easily bypass local authentication if no data returns from the authentication process.
```
Comment (sushi2k) [inline Document/0x06f-Testing-Local-Authentication.md]: same with 0x04e

File: Document/0x06f-Testing-Local-Authentication.md [hunk 3]
```diff
@@ -225,7 +229,8 @@ If vulnerable, the module will automatically bypass the login form.
 
 #### OWASP MASVS
 
-- V4.7: "Biometric authentication, if any, is not event-bound (i.e. using an API that simply returns "true" or "false"). Instead, it is based on unlocking the keychain/keystore."
+- V4.8: "Biometric authentication, if any, is not event-bound (i.e. using an API that simply returns "true" or "false"). Instead, it is based on unlocking the keychain/keystore."
+- 2.11: "The app enforces a minimum device-access-security policy, such as requiring the user to set a device passcode."
```
Comment (sushi2k) [inline Document/0x06f-Testing-Local-Authentication.md]: somehow the requirements are bold?

File: Document/0x04e-Testing-Authentication-and-Session-Management.md [hunk 4]
```diff
@@ -495,6 +497,21 @@ For additional best practices and detailed information please refer to the follo
 - [DRAFT - OAuth 2.0 for Native Apps](https://tools.ietf.org/html/draft-ietf-oauth-native-apps-12 "draft_ietf-oauth-native-apps-12: OAuth 2.0 for Native Apps (June 2017)")
 - [RFC6819 - OAuth 2.0 Threat Model and Security Considerations](https://tools.ietf.org/html/rfc6819 "RFC6819: OAuth 2.0 Threat Model and Security Considerations (January 2013)")
 
+### Login Activity and Device Blocking
+
+For applications which require L2 protection, the MASVS states that: "The app informs the user of all login activities with their account. Users are able view a list of devices used to access the account, and to block specific devices.". This can take part in various scenaios:
```
[the remark this answers is not in the document]
Comment (commjoen) [reply Document/0x04e-Testing-Authentication-and-Session-Management.md]: Fixed

File: Document/0x04e-Testing-Authentication-and-Session-Management.md [hunk 5]
```diff
@@ -495,6 +497,21 @@ For additional best practices and detailed information please refer to the follo
...
+2. The application provides an overview of the last session after login, if the previous session was with a different configuration (e.g. location, device, app-version) then the user his current configuration. The user then has the option to report suspicious activities and block devices used in the previous session.
+3. The application provides an overview of the last session after login at all times.
+4. The application has a self-service portal in which the user can see an audit-log and manage the different devices with which he can login.
+
+In all cases, the pentester should verify whether different devices are detected correctly. Therefore, the binding of the application to the actual device should be tested. For instance: in iOS a developer can use `identifierForVendor` whereas in Android, the developer can`Settings.Secure.ANDROID_ID` to identify an application instance. This togeter iwth keying material in the `Keychain` for iOS and in the `KeyStore` in Android can reassure strong device binding. Next, a pentester should test if using different IPs, different locations and/or different time-slots will trigger the right type of information in all scenarios.
```
[the remark this answers is not in the document]
Comment (commjoen) [reply Document/0x04e-Testing-Authentication-and-Session-Management.md]: Fixed both

File: Document/0x05f-Testing-Local-Authentication.md [hunk 6]
```diff
@@ -1,6 +1,100 @@
...
+During local authentication, an app authenticates the user against credentials stored locally on the device. In other words, the user "unlocks" the app or some inner layer of functionality by providing a valid PIN, password, or fingerprint, verified by referencing local data. Generally, this process is invoked for reasons such providing a user convenience for resuming an existing session with the remote service or as a means of step-up authentication to protect some critical function. 
+As described earlier in 0x04e: it is important to reassure that authentication happens at least on a cryptographic primitve (e.g.: an authentication step which results in unlocking a key). Next, it is recommended that the authentication is verified at a remote endpoint.
```
[the remark this answers is not in the document]
Comment (commjoen) [reply Document/0x05f-Testing-Local-Authentication.md]: Fxied

File: Document/0x05f-Testing-Local-Authentication.md [hunk 7]
```diff
@@ -1,6 +1,100 @@
...
+In Android, there are two mechanisms supported by the Android Runtime for local authentication: the Confirm Credential flow and the Biometric Authentication flow.
+
+
+### Testing Confirm Credentials
+
+#### Overview
+The confirm credential flow is avaiable since Android 6.0 and is used to ensure that users do not have to enter app-specific passwords together with the lockscreen-protection. Instead: if a user has logged in to his device recently, then confirm-credentials can be used to unlock cryptographic materials from the `AndroidKeystore`. That is, if the user unlocked his device within the set time limits (`setUserAuthenticationValidityDurationSeconds`), otherwise he has to unlock his device again.
```
[the remark this answers is not in the document]
Comment (commjoen) [reply Document/0x05f-Testing-Local-Authentication.md]: Fixed

File: Document/0x05f-Testing-Local-Authentication.md (hunk 1 above)
Comment (commjoen) [reply Document/0x05f-Testing-Local-Authentication.md]: Merged them

File: Document/0x06f-Testing-Local-Authentication.md (hunk 2 above)
Comment (commjoen) [reply Document/0x06f-Testing-Local-Authentication.md]: Fixed

File: Document/0x06f-Testing-Local-Authentication.md (hunk 3 above)
Comment (commjoen) [reply Document/0x06f-Testing-Local-Authentication.md]: this is because of the __bridge part. now escaped it looks better i guess...

Comment (commjoen) [reply Document/0x06f-Testing-Local-Authentication.md]: Should be fixed now, was basically due to the usage of __

Comment (commjoen) [reply Document/0x06f-Testing-Local-Authentication.md]: So using objc instead of objective-c fixed it in the end... it ws a parser issue

**Comment by commjoen** [comment]
All review comments fixed, ready for another round

**Comment by sushi2k** [comment]
Nice. Thanks Jeroen. Just merged it 👍

**Pull request merged**